ops/charm.py

```diff
@@ -1417,6 +1423,103 @@ def config(self) -> model.ConfigData:
         """A mapping containing the charm's config and current values."""
         return self.model.config
 
+    def load_config(
+        self,
+        cls: type[_ConfigType],
+        *args: Any,
+        errors: Literal['raise', 'blocked'] = 'raise',
+        **kwargs: Any,
+    ) -> _ConfigType:
+        """Load the config into an instance of a config class.
+
+        The raw Juju config is passed to the config class's ``__init__``, as
+        keyword arguments, with the following changes:
+
+        * ``secret`` type options have a :class:`model.Secret` value rather
+          than the secret ID. Note that the secret object is not validated by
+          Juju at this time, so may raise :class:`SecretNotFoundError` when it
+          is later used if the secret does not exist or the unit does not have
+          permission to access it.
+        * dashes in names converted to underscores.
+
+        Any additional positional or keyword arguments to this method will be
+        passed through to the config class ``__init__``.
+
+        Args:
+            cls: A class that will accept the Juju options as keyword arguments.
+            errors: what to do if the config is invalid. If ``blocked``, the
+                charm will exit successfully (this informs Juju that
+                the event was handled and it will not be retried) after setting
+                an appropriate blocked status. If ``raise``, ``load_config``
+                will not catch any exceptions, leaving the charm to handle
+                errors.
+            args: positional arguments to pass through to the config class.
+            kwargs: keyword arguments to pass through to the config class.
+
+        Returns:
+            An instance of the config class with the current config values.
+
+        Raises:
+            ValueError: if the configuration is invalid and ``errors`` is set to
+                ``raise``.
+        """
+        from ._main import _Abort
+
+        config: dict[str, bool | int | float | str | model.Secret] = kwargs.copy()
+        fields = set(_juju_option_names(cls))
+        for key, value in self.config.items():
+            attr = key.replace('-', '_')
+            if not attr.isidentifier():
+                if errors == 'raise':
+                    raise ValueError(f'Invalid attribute name {attr}')
+                self.unit.status = model.BlockedStatus(f'Invalid attribute name {attr}')
+                raise _Abort(0)
```

Comment:
When I hear back from the Juju team about what form an option can take I'll either remove this or add a comment explaining why it's needed.

Reply:
No response, but I've tested things out myself and it seems like anything that's both a valid Go string and a valid YAML field name is fine. There are lots of things that work fine in Juju but are not valid Python identifiers, like:

* `t$`
* `class`
* `a b`
* `2`
* `1a`

These are all fine as dictionary keys (as in `Model.config`) but not really as class/object attribute names. You can do a `**kwargs` and `setattr` to get them into the `__dict__` but it's pretty weird.

However, I guess you already know that you've chosen inappropriate config names when you write the class, and this was predominately there to protect against bad naming functions, which we don't provide any more. So I'll remove it.